src/Uart/Uart.c

```diff
@@ -0,0 +1,441 @@
+    }
+#ifndef RTEMS_SIS
+    switch(config->parity) {
+        case Uart_Parity_Even:
+            uart->reg->control |= UART_CONTROL_PE;
+            uart->reg->control &= ~UART_CONTROL_PS;
+            break;
+        case Uart_Parity_Odd:
+            uart->reg->control |= (UART_CONTROL_PE | UART_CONTROL_PS);
+            break;
+        case Uart_Parity_None:
+        default:
+            uart->reg->control &= ~UART_CONTROL_PE;
+            break;
+    }
+#endif
+}
+
+void
+Uart_getConfig(const Uart* const uart, Uart_Config* const config)
+{
+    config->isTxEnabled = (uart->reg->control & UART_CONTROL_TE) != 0;
+    config->isRxEnabled = (uart->reg->control & UART_CONTROL_RE) != 0;
+    config->isTestModeEnabled = (uart->reg->control & UART_CONTROL_LB) != 0;
+#ifndef RTEMS_SIS
+    if((uart->reg->control & UART_CONTROL_PE) == 0) {
+        config->parity = Uart_Parity_None;
+    } else if((uart->reg->control & UART_CONTROL_PS) != 0) {
+        config->parity = Uart_Parity_Odd;
+    } else {
+        config->parity = Uart_Parity_Even;
+    }
+    config->baudRate = getBaudRate(uart);
+#endif
+    config->baudRateClkSrc = GPTIMER_ADDRESS_BASE;
+    config->baudRateClkFreq = rtems_clock_get_ticks_per_second();
+}
+
+void
+Uart_startup(Uart* const uart)
+{
+    sentBytesCounter[uart->id] = 0;
+    errorFlags[uart->id] = (Uart_ErrorFlags){ 0 };
+    rtems_interrupt_entry_initialize(
+            &(uartIrqEntry[uart->id]),
+            (rtems_interrupt_handler)Uart_handleInterrupt,
+            uart,
+            "Uart Interrupt");
+    rtems_interrupt_entry_install(interruptNumber(uart->id),
+                                  RTEMS_INTERRUPT_UNIQUE,
+                                  &(uartIrqEntry[uart->id]));
+    rtems_interrupt_vector_enable(interruptNumber(uart->id));
+}
+
+void
+Uart_shutdown(Uart* const uart)
+{
+    rtems_interrupt_vector_disable(interruptNumber(uart->id));
+    rtems_interrupt_entry_remove(interruptNumber(uart->id),
+                                 &(uartIrqEntry[uart->id]));
+    uart->reg->control = 0;
+    uart->reg->status = 0;
+    sentBytesCounter[uart->id] = 0;
+    errorFlags[uart->id] = (Uart_ErrorFlags){ 0 };
+}
+
+void
+Uart_init(Uart_Id id, Uart* const uart)
+{
+    uart->id = id;
+    uart->reg = getAddressBase(id);
+    sentBytesCounter[uart->id] = 0;
+    errorFlags[uart->id] = (Uart_ErrorFlags){ 0 };
+    uart->txHandler = defaultTxHandler;
+    uart->rxHandler = defaultRxHandler;
+    uart->errorHandler = defaultErrorHandler;
+    uart->txFifo = NULL;
+    uart->rxFifo = NULL;
+    Uart_shutdown(uart);
+    rtems_interrupt_clear(interruptNumber(uart->id));
+}
+
+bool
+Uart_write(Uart* const uart,
+           const uint8_t data,
+           uint32_t const timeoutLimit,
+           Uart_ErrorCode* const errCode)
+{
+    rtems_interrupt_vector_disable(interruptNumber(uart->id));
+    uint32_t timeout = timeoutLimit;
+    if(uart->txFifo == NULL) {
+        if((uart->reg->status & UART_STATUS_TE) != 0) {
```

Comment:
I guess that here we check whether the transmitter is enabled, and later at line 240, we check in the loop if data was sent. 
What I am missing here is an "else", which would return immediately if the transmitter is not enabled, instead of waiting for a timeout.

Reply:
UART_STATUS_TE is 1 when data register in Uart device is empty, ie. tramsitter is ready to start sending byte of data. It shouldn't fail of it's 0, just wait until device is ready. In my opinion if device is not ready in specified period of time then timeout should occur, rather than failing immediately.